File: app/schemas/games.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator


class WheelSegment(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True, extra="forbid")
    title: str = Field(min_length=1, max_length=100)
    weight: int = Field(ge=1, le=10000)
    xp: int = Field(default=0, ge=0, le=10000)
    coins: int = Field(default=0, ge=0, le=1000)


class WheelInput(BaseModel):
    model_config = ConfigDict(extra="forbid")
    enabled: bool
    daily_spins: int = Field(ge=1, le=10)
    segments: list[WheelSegment] = Field(min_length=2, max_length=12)


class SpinInput(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)
    request_id: str = Field(min_length=16, max_length=80)


class RaffleInput(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)
    title: str = Field(min_length=1, max_length=180)
    description: str = Field(default="", max_length=4000)
    prize: str = Field(min_length=1, max_length=180)
    closes_at: datetime
    status: Literal["draft", "published"] = "draft"
    xp_reward: int = Field(default=0, ge=0, le=100000)
    coins_reward: int = Field(default=0, ge=0, le=10000)

    @model_validator(mode="after")
    def aware_deadline(self):
        if self.closes_at.tzinfo is None:
            raise ValueError("Укажите часовой пояс срока розыгрыша")
        return self

File: frontend/src/api/games.ts
import { request } from "./client";
import type { Page } from "./types";
export interface WheelSegment { title: string; weight: number; xp: number; coins: number }
export interface WheelConfig { enabled: boolean; daily_spins: number; segments: WheelSegment[]; used_today: number }
export interface WheelSpin { id: number; segment: number; reward: WheelSegment; segments: WheelSegment[]; created_at: string }
export interface Raffle { id: number; title: string; description: string; prize: string; closes_at: string; status: "draft" | "published" | "closed"; xp_reward: number; coins_reward: number; participants: number; entered: boolean; winner_name: string | null; won: boolean; drawn_at: string | null; accepting_entries: boolean }
export type RaffleInput = Pick<Raffle,"title" | "description" | "prize" | "closes_at" | "xp_reward" | "coins_reward"> & { status: "draft" | "published" };
export const games = {
  wheel: () => request<WheelConfig>("/api/v1/games/wheel"),
  configureWheel: (config: Omit<WheelConfig,"used_today">) => request<WheelConfig>("/api/v1/admin/games/wheel", { method: "PUT", json: config }),
  spin: (request_id: string) => request<WheelSpin>("/api/v1/games/wheel/spin", { method: "POST", json: { request_id } }),
  history: (page: number) => request<Page<WheelSpin>>(`/api/v1/games/wheel/history?page=${page}`),
  raffles: (admin = false) => request<Raffle[]>(`/api/v1/${admin ? "admin/" : ""}games/raffles`),
  saveRaffle: (data: RaffleInput, id?: number) => request<Raffle>(`/api/v1/admin/games/raffles${id ? `/${id}` : ""}`, { method: id ? "PUT" : "POST", json: data }),
  enter: (id: number) => request<{ detail: string }>(`/api/v1/games/raffles/${id}/enter`, { method: "POST" }),
  draw: (id: number) => request<Raffle>(`/api/v1/admin/games/raffles/${id}/draw`, { method: "POST" }),
};
